### PR TITLE
2772 match annotations in pydantic schema

### PR DESCRIPTION
**Pull request by nicholasbutlin**
## Description

Changed the implementation of the strawberry inference of type annotations to make sure it matches the pydantic annotations. I.e. correctly returning if it is Optional or Not ,whilst also ensuring it uses strawberry compatible types 

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #2772 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Nick Butlin for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/experimental/pydantic/object_type.py [hunk 1]
```diff
@@ -43,11 +43,7 @@
 def get_type_for_field(field: ModelField, is_input: bool):  # noqa: ANN201
     outer_type = field.outer_type_
     replaced_type = replace_types_recursively(outer_type, is_input)
-
-    default_defined: bool = (
-        field.default_factory is not None or field.default is not None
-    )
-    should_add_optional: bool = not (field.required or default_defined)
+    should_add_optional: bool = field.allow_none
```
Comment (thejaminator) [inline strawberry/experimental/pydantic/object_type.py:46]: oh cool!

**Comment by thejaminator** [comment]
thanks a lot @nicholasbutlin ! this looks great. pinging @patrick91 just to double check the release notes if they are ok

**Review by patrick91 — APPROVED** [review]
Awesome, thanks!

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**